src/lib/OS/SimOS.cc

```diff
@@ -263,6 +262,91 @@ uint64_t SimOS::createProcess(span<char> instructionBytes) {
   return tid;
 }
 
+int64_t SimOS::cloneProcess(uint64_t flags, uint64_t stackPtr,
+                            uint64_t parentTidPtr, uint64_t tls,
+                            uint64_t childTidPtr, uint64_t parentTid,
+                            uint64_t coreID, Register retReg) {
+  /** Supported Flags :
+   * CLONE_VM | CLONE_FS | CLONE_FILES | CLONE_THREAD | CLONE_SYSVSEM |
+   * CLONE_SETTLS| CLONE_CHILD_CLEARTID | CLONE_SIGHAND | CLONE_PARENT_SETTID */
+
+  // Get TGID of calling process : as we require the CLONE_THREAD flag the new
+  // thread will be in the same thread group. Given we copy the parent process
+  // object however, the TGID will match to parent implicitly
+  uint64_t newChildTid = nextFreeTID_;
+  nextFreeTID_++;
+  // Ignore CLONE_SIGHAND flag for now
+  // Ignore CLONE_SYSVSEM flag for now
+  // Clone from parent Process object
+  std::shared_ptr<Process> newProc =
+      std::make_shared<Process>(*processes_[parentTid]);
+
+  // Update TID
+  newProc->updateTID(newChildTid);
+
+  // Update childTidPtr value. Defaults to 0 if CLONE_CHILD_CLEARTID is not
+  // present
+  newProc->clearChildTid_ = childTidPtr;
+
+  // Update stackPtr, stackSize and stackEnd
+  newProc->updateStack(stackPtr);
+
+  // TLS (Thread Local Storage) region already mapped
+
+  // Store child tid at parentTidPtr if required
+  uint64_t paddr = handleVAddrTranslation(parentTidPtr, parentTid);
+  if (masks::faults::hasFault(paddr)) {
+    std::cout << "Fault in parentTidPtr translation in cloneProcess"
+              << std::endl;
+    std::exit(1);
+  }
+  if (flags & f_CLONE_PARENT_SETTID) {
+    std::vector<char> dataVec(sizeof(newChildTid), '\0');
+    std::memcpy(dataVec.data(), &newChildTid, sizeof(newChildTid));
+    memory_->sendUntimedData(dataVec, paddr, sizeof(newChildTid));
+  }
+
+  // Update context of new child process to match parent process's current state
+  // in the Core, updating the TID
+  cpuContext currContext = cores_[coreID]->getCurrentContext();
+  newProc->context_.pc = currContext.pc;
+  newProc->context_.regFile = currContext.regFile;
+  // Update returnRegister value to 0
+  newProc->context_.regFile[retReg.type][retReg.tag] = {0, 8};
+  // Update stack pointer
+  newProc->context_.sp = stackPtr;
+  if (Config::get()["Core"]["ISA"].as<std::string>() == "rv64") {
+    newProc->context_.regFile[arch::riscv::RegisterType::GENERAL][2] = {
+        stackPtr, 8};
+  } else if (Config::get()["Core"]["ISA"].as<std::string>() == "AArch64") {
+    newProc->context_.regFile[arch::aarch64::RegisterType::GENERAL][31] = {
+        stackPtr, 8};
+    newProc->context_.regFile[arch::aarch64::RegisterType::SYSTEM]
+                             [arch::aarch64::ARM64_SYSREG_TAGS::TPIDR_EL0] = {
+        tls, 8};
+  }
+  newProc->context_.TID = newChildTid;
+  newProc->status_ = procStatus::waiting;
+  processes_.emplace(newChildTid, newProc);
+  waitingProcs_.push(newProc);
+
+  // Add stack to `proc/tgid/maps`
```

Comment:
I thought the writing in `Process.cc` added the mappings for stack and heap?

Reply:
It does yes, but the stack region for a child thread (i.e. after a `clone` system call) is created through a prior mmap call. As such, the new child's stack needs adding to its maps file.